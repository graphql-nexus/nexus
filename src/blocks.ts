export { InputDefinitionBlock, OutputDefinitionBlock } from './definitions/definitionBlocks'
export { InterfaceDefinitionBlock } from './definitions/interfaceType'
export { ObjectDefinitionBlock } from './definitions/objectType'
export { UnionDefinitionBlock } from './definitions/unionType'
