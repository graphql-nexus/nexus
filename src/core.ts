// The "core" is used as a namespace to re-export everything,
// For anyone who wants to use the internals
export * from './builder'
export * from './makeSchema'
export * from './definitions/args'
export * from './definitions/decorateType'
export * from './definitions/definitionBlocks'
export * from './definitions/enumType'
export * from './definitions/extendInputType'
export * from './definitions/extendType'
export * from './definitions/inputObjectType'
export * from './definitions/interfaceType'
export * from './definitions/list'
export * from './definitions/mutationField'
export * from './definitions/mutationType'
export * from './definitions/nonNull'
export * from './definitions/nullable'
export * from './definitions/objectType'
export * from './definitions/queryField'
export * from './definitions/queryType'
export * from './definitions/scalarType'
export * from './definitions/subscriptionField'
export * from './definitions/subscriptionType'
export * from './definitions/nexusMeta'
export * from './definitions/unionType'
export * from './definitions/wrapping'
export * from './definitions/_types'
export * from './dynamicMethod'
export * from './plugin'
export * from './plugins'
export * from './rebuildType'
export * from './sdlConverter'
export * from './typegenAutoConfig'
export * from './typegenFormatPrettier'
export * from './typegenMetadata'
export * from './typegenPrinter'
export * from './typegenTypeHelpers'
export * from './typegenUtils'
export * from './utils'
