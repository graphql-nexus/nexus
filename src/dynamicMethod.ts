import { PluginBuilderLens } from './builder'
import { InputDefinitionBlock, OutputDefinitionBlock } from './definitions/definitionBlocks'
import { NexusTypes, withNexusSymbol } from './definitions/_types'

export type OutputFactoryConfig<T> = {
  stage: 'walk' | 'build'
  args: any[]
  builder: PluginBuilderLens
  typeDef: OutputDefinitionBlock<any>
  /**
   * The name of the type this field is being declared on
   */
  typeName: string
}

export type InputFactoryConfig<T> = {
  args: any[]
  builder: PluginBuilderLens
  typeDef: InputDefinitionBlock<any>
  /**
   * The name of the type this field is being declared on
   */
  typeName: string
}

export interface BaseExtensionConfig<T extends string> {
  /**
   * The name of the "extension", the field made
   * available on the builders
   */
  name: T
  /**
   * The full type definition for the options, including generic
   * signature for the type
   */
  typeDefinition?: string
}

export interface DynamicOutputMethodConfig<T extends string> extends BaseExtensionConfig<T> {
  /**
   * Invoked when the field is called
   */
  factory(config: OutputFactoryConfig<T>): any
}

export interface DynamicInputMethodConfig<T extends string> extends BaseExtensionConfig<T> {
  /**
   * Invoked when the field is called
   */
  factory(config: InputFactoryConfig<T>): any
}

export class DynamicInputMethodDef<Name extends string> {
  constructor(readonly name: Name, protected config: DynamicInputMethodConfig<Name>) {}
  get value() {
    return this.config
  }
}
withNexusSymbol(DynamicInputMethodDef, NexusTypes.DynamicInput)

export class DynamicOutputMethodDef<Name extends string> {
  constructor(readonly name: Name, protected config: DynamicOutputMethodConfig<Name>) {}
  get value() {
    return this.config
  }
}
withNexusSymbol(DynamicOutputMethodDef, NexusTypes.DynamicOutputMethod)

/**
 * Defines a new property on the object definition block
 * for an output type, taking arbitrary input to define
 * additional types. See the connectionPlugin:
 *
 * t.connectionField('posts', {
 *   nullable: true,
 *   totalCount(root, args, ctx, info) {
 *     return ctx.user.getTotalPostCount(root.id, args)
 *   },
 *   nodes(root, args, ctx, info) {
 *     return ctx.user.getPosts(root.id, args)
 *   }
 * })
 */
export function dynamicOutputMethod<T extends string>(config: DynamicOutputMethodConfig<T>) {
  return new DynamicOutputMethodDef(config.name, config)
}

/**
 * Same as the outputFieldExtension, but for fields that
 * should be added on as input types.
 */
export function dynamicInputMethod<T extends string>(config: DynamicInputMethodConfig<T>) {
  return new DynamicInputMethodDef(config.name, config)
}
