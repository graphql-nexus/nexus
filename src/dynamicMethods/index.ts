export * from './collection'
export * from './relayConnection'
