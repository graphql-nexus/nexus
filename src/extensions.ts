import { GraphQLNamedType } from 'graphql'
import { DynamicFieldDefs, SchemaConfig } from './builder'
import { NexusOutputFieldConfig } from './definitions/definitionBlocks'
import { NexusInputObjectTypeConfig } from './definitions/inputObjectType'
import { NexusInterfaceTypeConfig } from './definitions/interfaceType'
import { NexusObjectTypeConfig } from './definitions/objectType'
import { RootTypings } from './definitions/_types'

export type NexusGraphQLNamedType = GraphQLNamedType & {
  extensions?: {
    nexus?: {
      config: any
    }
  }
}

export type NexusTypeExtensions = NexusObjectTypeExtension | NexusInterfaceTypeExtension

/**
 * Container object living on `fieldDefinition.extensions.nexus`
 */
export class NexusFieldExtension<TypeName extends string = any, FieldName extends string = any> {
  readonly config: Omit<NexusOutputFieldConfig<TypeName, FieldName>, 'resolve'>
  /**
   * Whether the user has provided a custom "resolve" function,
   * or whether we're using GraphQL's defaultResolver
   */
  readonly hasDefinedResolver: boolean
  constructor(config: NexusOutputFieldConfig<TypeName, FieldName>) {
    const { resolve, ...rest } = config
    this.config = rest
    this.hasDefinedResolver = Boolean(resolve)
  }
}

/**
 * Container object living on `inputObjectType.extensions.nexus`
 */
export class NexusInputObjectTypeExtension<TypeName extends string = any> {
  readonly config: Omit<NexusInputObjectTypeConfig<TypeName>, 'definition'>
  constructor(config: NexusInputObjectTypeConfig<TypeName>) {
    const { definition, ...rest } = config
    this.config = rest
  }
}

/**
 * Container object living on `objectType.extensions.nexus`
 */
export class NexusObjectTypeExtension<TypeName extends string = any> {
  readonly config: Omit<NexusObjectTypeConfig<TypeName>, 'definition'>
  constructor(config: NexusObjectTypeConfig<TypeName>) {
    const { definition, ...rest } = config
    this.config = rest
  }
}

/**
 * Container object living on `interfaceType.extensions.nexus`
 */
export class NexusInterfaceTypeExtension<TypeName extends string = any> {
  readonly config: Omit<NexusInterfaceTypeConfig<TypeName>, 'definition'>
  constructor(config: NexusInterfaceTypeConfig<TypeName>) {
    const { definition, ...rest } = config
    this.config = rest
  }
}

export interface NexusSchemaExtensionConfig extends Omit<SchemaConfig, 'types'> {
  dynamicFields: DynamicFieldDefs
  rootTypings: RootTypings
}

/**
 * Container object living on `schema.extensions.nexus`. Keeps track
 * of metadata from the builder so we can use it when we
 */
export class NexusSchemaExtension {
  constructor(readonly config: NexusSchemaExtensionConfig) {}
}
