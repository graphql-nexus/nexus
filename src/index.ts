import * as blocks from './blocks'
import * as core from './core'

// All of the Public API definitions
export { makeSchema } from './makeSchema'
export { arg, booleanArg, floatArg, idArg, intArg, stringArg } from './definitions/args'
export { decorateType } from './definitions/decorateType'
export { enumType } from './definitions/enumType'
export { extendInputType } from './definitions/extendInputType'
export { extendType } from './definitions/extendType'
export { inputObjectType } from './definitions/inputObjectType'
export { interfaceType } from './definitions/interfaceType'
export { list } from './definitions/list'
export { mutationField } from './definitions/mutationField'
export { mutationType } from './definitions/mutationType'
export { nonNull } from './definitions/nonNull'
export { nullable } from './definitions/nullable'
export { objectType } from './definitions/objectType'
export { queryField } from './definitions/queryField'
export { queryType } from './definitions/queryType'
export { asNexusMethod, scalarType } from './definitions/scalarType'
export { subscriptionField } from './definitions/subscriptionField'
export { subscriptionType } from './definitions/subscriptionType'
export { unionType } from './definitions/unionType'
export { dynamicInputMethod, dynamicOutputMethod } from './dynamicMethod'
export { dynamicOutputProperty } from './dynamicProperty'
export { createPlugin, plugin, PluginBuilderLens, PluginConfig } from './plugin'
export * from './plugins'
export { convertSDL } from './sdlConverter'
export { AllInputTypes, AllOutputTypes, FieldResolver, FieldType } from './typegenTypeHelpers'
export { groupTypes } from './utils'
export { blocks, core }
