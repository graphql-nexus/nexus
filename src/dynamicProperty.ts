import { SchemaBuilder } from './builder'
import { OutputDefinitionBlock } from './definitions/definitionBlocks'
import { NexusTypes, withNexusSymbol } from './definitions/_types'
import { BaseExtensionConfig } from './dynamicMethod'

export type OutputPropertyFactoryConfig<T> = {
  stage: 'walk' | 'build'
  builder: SchemaBuilder
  typeDef: OutputDefinitionBlock<any>
  /**
   * The name of the type this field is being declared on
   */
  typeName: string
}

export interface DynamicOutputPropertyConfig<T extends string> extends BaseExtensionConfig<T> {
  /**
   * Invoked when the property is accessed (as a getter)
   */
  factory(config: OutputPropertyFactoryConfig<T>): any
}

export class DynamicOutputPropertyDef<Name extends string> {
  constructor(readonly name: Name, protected config: DynamicOutputPropertyConfig<Name>) {}
  get value() {
    return this.config
  }
}
withNexusSymbol(DynamicOutputPropertyDef, NexusTypes.DynamicOutputProperty)

/**
 * Defines a new property on the object definition block
 * for an output type, making it possible to build custom DSL's
 * on top of Nexus, e.g. in nexus-prisma
 *
 * t.model.posts()
 */
export function dynamicOutputProperty<T extends string>(config: DynamicOutputPropertyConfig<T>) {
  return new DynamicOutputPropertyDef(config.name, config)
}
