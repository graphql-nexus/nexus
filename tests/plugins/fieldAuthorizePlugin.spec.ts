import { graphql } from 'graphql'
import path from 'path'
import { fieldAuthorizePlugin, makeSchema, objectType, queryField } from '../../src'
import { generateSchema, declarativeWrappingPlugin } from '../../src/core'

describe('fieldAuthorizePlugin', () => {
  const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => {})
  const consoleWarnSpy = jest.spyOn(console, 'warn').mockImplementation(() => {})

  afterEach(() => {
    jest.resetAllMocks()
  })

  const schemaTypes = [
    objectType({
      name: 'User',
      definition(t) {
        t.int('id')
      },
    }),
    queryField('userTrue', {
      type: 'User',
      // @ts-ignore
      authorize: (root, args, ctx) => ctx.user.id === 1,
      resolve: () => ({ id: 1 }),
    }),
    queryField('userTruePromise', {
      type: 'User',
      // @ts-ignore
      authorize: async (root, args, ctx) => ctx.user.id === 1,
      resolve: () => ({ id: 1 }),
    }),
    queryField('userFalse', {
      type: 'User',
      // @ts-ignore
      authorize: (root, args, ctx) => ctx.user.id === 2,
      resolve: () => ({ id: 2 }),
    }),
    queryField('userFalsePromise', {
      type: 'User',
      // @ts-ignore
      authorize: async (root, args, ctx) => ctx.user.id === 2,
      resolve: () => ({ id: 2 }),
    }),
    queryField('userReturnError', {
      type: 'User',
      // @ts-ignore
      authorize: (root, args, ctx) => new Error('You shall not pass.'),
      resolve: () => ({ id: 1 }),
    }),
    queryField('userReturnPromiseError', {
      type: 'User',
      // @ts-ignore
      authorize: (root, args, ctx) => new Error('You shall not pass.'),
      resolve: () => ({ id: 1 }),
    }),
    queryField('userThrowError', {
      type: 'User',
      // @ts-ignore
      authorize: (root, args, ctx) => {
        throw new Error('You shall not pass.')
      },
      resolve: () => ({ id: 1 }),
    }),
    queryField('userInvalidValue', {
      type: 'User',
      // @ts-ignore
      authorize: (root, args, ctx) => 1,
      resolve: () => ({ id: 1 }),
    }),
  ]
  const testSchema = makeSchema({
    outputs: false,
    types: schemaTypes,
    nonNullDefaults: {
      output: true,
    },
    plugins: [
      fieldAuthorizePlugin({
        formatError({ error, root, args, ctx, info }) {
          console.error(`Guarded error ${error.message}`)
          return new Error('Authorization Error')
        },
      }),
    ],
  })
  const mockCtx = { user: { id: 1 } }
  const testField = (field: string, passes = false, schema = testSchema) => {
    return graphql(
      schema,
      `
        {
          ${field} {
            id
          }
        }
      `,
      {},
      mockCtx
    )
  }

  test('field-level authorize passes returning true', async () => {
    const { data, errors = [] } = await testField('userTrue', true)
    expect(consoleErrorSpy).not.toHaveBeenCalled()
    expect(errors).toEqual([])
    expect(data?.userTrue).toEqual({ id: 1 })
  })

  test('field-level authorize passes returning a Promise for true', async () => {
    const { data, errors = [] } = await testField('userTruePromise', true)
    expect(consoleErrorSpy).not.toHaveBeenCalled()
    expect(errors).toEqual([])
    expect(data?.userTruePromise).toEqual({ id: 1 })
  })

  test('field-level authorize fails returning false', async () => {
    const { data, errors = [] } = await testField('userFalse')
    expect(data).toBeNull()
    expect(errors.length).toEqual(1)
    expect(errors[0].message).toEqual('Authorization Error')
    expect(consoleErrorSpy).toHaveBeenCalledTimes(1)
    expect(consoleErrorSpy.mock.calls[0]).toMatchSnapshot()
  })

  test('field-level authorize fails returning a Promise for false', async () => {
    const { data, errors = [] } = await testField('userFalsePromise')
    expect(data).toBeNull()
    expect(errors.length).toEqual(1)
    expect(errors[0].message).toEqual('Authorization Error')
    expect(consoleErrorSpy).toHaveBeenCalledTimes(1)
    expect(consoleErrorSpy.mock.calls[0]).toMatchSnapshot()
  })

  test('field-level authorize fails returning an error', async () => {
    const { data, errors = [] } = await testField('userReturnError')
    expect(data).toBeNull()
    expect(errors.length).toEqual(1)
    expect(errors[0].message).toEqual('Authorization Error')
    expect(consoleErrorSpy).toHaveBeenCalledTimes(1)
    expect(consoleErrorSpy.mock.calls[0]).toMatchSnapshot()
  })

  test('field-level authorize fails returning a Promise for an error', async () => {
    const { data, errors = [] } = await testField('userFalsePromise')
    expect(data).toBeNull()
    expect(errors.length).toEqual(1)
    expect(errors[0].message).toEqual('Authorization Error')
    expect(consoleErrorSpy).toHaveBeenCalledTimes(1)
    expect(consoleErrorSpy.mock.calls[0]).toMatchSnapshot()
  })

  test('field-level authorize fails throwing an error', async () => {
    const { data, errors = [] } = await testField('userThrowError')
    expect(data).toBeNull()
    expect(errors.length).toEqual(1)
    expect(errors[0].message).toEqual('Authorization Error')
    expect(consoleErrorSpy).toHaveBeenCalledTimes(1)
    expect(consoleErrorSpy.mock.calls[0]).toMatchSnapshot()
  })

  test('field-level authorize fails with any other value', async () => {
    const { data, errors = [] } = await testField('userInvalidValue')
    expect(data).toBeNull()
    expect(errors.length).toEqual(1)
    expect(errors[0].message).toEqual('Authorization Error')
    expect(consoleErrorSpy).toHaveBeenCalledTimes(1)
    expect(consoleErrorSpy.mock.calls[0]).toMatchSnapshot()
  })

  test('default format error does not call console.error', async () => {
    const schema = makeSchema({
      outputs: false,
      types: schemaTypes,
      nonNullDefaults: {
        output: true,
      },
      plugins: [fieldAuthorizePlugin()],
    })
    const { data, errors = [] } = await testField('userFalse', false, schema)
    expect(data).toBeNull()
    expect(errors.length).toEqual(1)
    expect(errors[0].message).toEqual('Not authorized')
    expect(consoleErrorSpy).toHaveBeenCalledTimes(0)
  })

  test('always throws an error, even when formatError is wrong', async () => {
    const schema = makeSchema({
      outputs: false,
      nonNullDefaults: {
        output: true,
      },
      types: schemaTypes,
      plugins: [
        // @ts-ignore
        fieldAuthorizePlugin({
          // @ts-ignore
          formatError() {},
        }),
      ],
    })
    const { data, errors = [] } = await testField('userFalse', false, schema)
    expect(data).toBeNull()
    expect(errors.length).toEqual(1)
    expect(errors[0].message).toEqual('Not authorized')
    expect(consoleErrorSpy).toHaveBeenCalledTimes(1)
    expect(consoleErrorSpy.mock.calls[0]).toMatchSnapshot()
  })

  test('warns when a field has a non-function authorize prop', async () => {
    const schema = makeSchema({
      outputs: false,
      types: [
        schemaTypes,
        queryField('incorrectFieldConfig', {
          type: 'User',
          // @ts-ignore
          authorize: 1,
          resolve: () => ({ id: 1 }),
        }),
      ],
      plugins: [fieldAuthorizePlugin({})],
    })
    const { data, errors = [] } = await testField('incorrectFieldConfig', true, schema)
    expect(errors).toEqual([])
    expect(data?.incorrectFieldConfig).toEqual({ id: 1 })
    expect(consoleErrorSpy).toHaveBeenCalledTimes(1)
    expect(consoleErrorSpy.mock.calls[0]).toMatchSnapshot()
  })

  test('warns and adds the authorize plugin when a schema has an authorize prop but no plugins', async () => {
    const schema = makeSchema({
      outputs: false,
      nonNullDefaults: {
        output: true,
      },
      types: [
        schemaTypes,
        queryField('shouldWarn', {
          type: 'User',
          // @ts-ignore
          authorize: () => true,
          resolve: () => ({ id: 1 }),
        }),
      ],
    })
    const { data, errors = [] } = await testField('shouldWarn', true, schema)
    expect(errors).toEqual([])
    expect(data?.shouldWarn).toEqual({ id: 1 })
    expect(consoleErrorSpy).toHaveBeenCalledTimes(0)
    expect(consoleWarnSpy).toHaveBeenCalledTimes(1)
    expect(consoleWarnSpy.mock.calls[0]).toMatchSnapshot()
  })

  test('printing the authorize schema', async () => {
    const result = await generateSchema.withArtifacts(
      {
        types: [
          queryField('ok', {
            type: 'Boolean',
            resolve: () => true,
          }),
        ],
        plugins: [fieldAuthorizePlugin(), declarativeWrappingPlugin({ disable: true })],
      },
      path.join(__dirname, 'test.gen.ts')
    )
    expect(result.tsTypes).toMatchSnapshot('Full Type Output')
  })
})
