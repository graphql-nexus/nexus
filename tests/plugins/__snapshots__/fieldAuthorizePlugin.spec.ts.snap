// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fieldAuthorizePlugin always throws an error, even when formatError is wrong 1`] = `
Array [
  "Non-Error value undefined returned from custom formatError in authorize plugin",
]
`;

exports[`fieldAuthorizePlugin field-level authorize fails returning a Promise for an error 1`] = `
Array [
  "Guarded error Not authorized",
]
`;

exports[`fieldAuthorizePlugin field-level authorize fails returning a Promise for false 1`] = `
Array [
  "Guarded error Not authorized",
]
`;

exports[`fieldAuthorizePlugin field-level authorize fails returning an error 1`] = `
Array [
  "Guarded error You shall not pass.",
]
`;

exports[`fieldAuthorizePlugin field-level authorize fails returning false 1`] = `
Array [
  "Guarded error Not authorized",
]
`;

exports[`fieldAuthorizePlugin field-level authorize fails throwing an error 1`] = `
Array [
  "Guarded error You shall not pass.",
]
`;

exports[`fieldAuthorizePlugin field-level authorize fails with any other value 1`] = `
Array [
  "Guarded error Nexus authorize for Query.userInvalidValue Expected a boolean or Error, saw 1",
]
`;

exports[`fieldAuthorizePlugin printing the authorize schema: Full Type Output 1`] = `
"/**
 * This file was generated by Nexus Schema
 * Do not make changes to this file directly
 */


import { FieldAuthorizeResolver } from \\"@nexus/schema/dist/plugins/fieldAuthorizePlugin\\"




declare global {
  interface NexusGen extends NexusGenTypes {}
}

export interface NexusGenInputs {
}

export interface NexusGenEnums {
}

export interface NexusGenScalars {
  String: string
  Int: number
  Float: number
  Boolean: boolean
  ID: string
}

export interface NexusGenRootTypes {
  Query: {};
}

export interface NexusGenAllTypes extends NexusGenRootTypes {
  String: NexusGenScalars['String'];
  Int: NexusGenScalars['Int'];
  Float: NexusGenScalars['Float'];
  Boolean: NexusGenScalars['Boolean'];
  ID: NexusGenScalars['ID'];
}

export interface NexusGenFieldTypes {
  Query: { // field return type
    ok: boolean | null; // Boolean
  }
}

export interface NexusGenFieldTypeNames {
  Query: { // field return type name
    ok: 'Boolean'
  }
}

export interface NexusGenArgTypes {
}

export interface NexusGenAbstractTypeMembers {
}

export interface NexusGenInheritedFields {}

export type NexusGenObjectNames = \\"Query\\";

export type NexusGenInputNames = never;

export type NexusGenEnumNames = never;

export type NexusGenInterfaceNames = never;

export type NexusGenScalarNames = \\"Boolean\\" | \\"Float\\" | \\"ID\\" | \\"Int\\" | \\"String\\";

export type NexusGenUnionNames = never;

export type NexusGenObjectsUsingAbstractStrategyIsTypeOf = never;

export type NexusGenAbstractsUsingStrategyResolveType = never;

export type NexusGenFeaturesConfig = {
  abstractTypeStrategies: {
    isTypeOf: true
    resolveType: false
    __typename: false
  }
}

export interface NexusGenTypes {
  context: any;
  inputTypes: NexusGenInputs;
  rootTypes: NexusGenRootTypes;
  argTypes: NexusGenArgTypes;
  fieldTypes: NexusGenFieldTypes;
  fieldTypeNames: NexusGenFieldTypeNames;
  allTypes: NexusGenAllTypes;
  inheritedFields: NexusGenInheritedFields;
  objectNames: NexusGenObjectNames;
  inputNames: NexusGenInputNames;
  enumNames: NexusGenEnumNames;
  interfaceNames: NexusGenInterfaceNames;
  scalarNames: NexusGenScalarNames;
  unionNames: NexusGenUnionNames;
  allInputTypes: NexusGenTypes['inputNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['scalarNames'];
  allOutputTypes: NexusGenTypes['objectNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['unionNames'] | NexusGenTypes['interfaceNames'] | NexusGenTypes['scalarNames'];
  allNamedTypes: NexusGenTypes['allInputTypes'] | NexusGenTypes['allOutputTypes']
  abstractTypes: NexusGenTypes['interfaceNames'] | NexusGenTypes['unionNames'];
  abstractTypeMembers: NexusGenAbstractTypeMembers;
  objectsUsingAbstractStrategyIsTypeOf: NexusGenObjectsUsingAbstractStrategyIsTypeOf;
  abstractsUsingStrategyResolveType: NexusGenAbstractsUsingStrategyResolveType;
  features: NexusGenFeaturesConfig;
}


declare global {
  interface NexusGenPluginTypeConfig<TypeName extends string> {
  }
  interface NexusGenPluginFieldConfig<TypeName extends string, FieldName extends string> {
    /**
     * Authorization for an individual field. Returning \\"true\\"
     * or \\"Promise<true>\\" means the field can be accessed.
     * Returning \\"false\\" or \\"Promise<false>\\" will respond
     * with a \\"Not Authorized\\" error for the field.
     * Returning or throwing an error will also prevent the
     * resolver from executing.
     */
    authorize?: FieldAuthorizeResolver<TypeName, FieldName>
  }
  interface NexusGenPluginSchemaConfig {
  }
}"
`;

exports[`fieldAuthorizePlugin warns and adds the authorize plugin when a schema has an authorize prop but no plugins 1`] = `
Array [
  "The GraphQL Nexus \\"authorize\\" feature has been moved to a plugin, add [fieldAuthorizePlugin()] to your makeSchema plugin config to remove this warning.",
]
`;

exports[`fieldAuthorizePlugin warns when a field has a non-function authorize prop 1`] = `
Array [
  [Error: The authorize property provided to incorrectFieldConfig with type User should be a function, saw number],
]
`;
